rapid7/metasploit-framework
# Add History Manager and separate history files for Pry and Meterpreter

## Note
This PR is done in collaboration with @smcintyre-r7. Anyone can check his awesome work [here](https://github.com/pingport80/metasploit-framework/pull/1).
## Summary
This PR adds a `history manager` which can be used for tracking histories of various sub-shells separately. This adds separate history files for `Meterpreter` and `Pry` and fixes #10149.

Earlier the history file for Pry was `~/.pry_history` which is used system-wide and also pushes commands which were used outside metasploit for some other reason. This adds a `msf` specific history file for pry in `~/.msf4/pry_history`.

The `history manager` can also be used in future for other sub-shells for tracking their histories. Even if some issues arise because of some specific sub-shell this will make sure that history of `msfconsole` doesn't get messed up like in #10149.

Thanks to @acammack-r7, @gwillcox-r7  and @zeroSteiner for their suggestions :).

### Verification Steps For Meterpreter history
Switch between `meterpreter`, `msfconsole` and `pry` and check for histories.
```
msf6 > history 
1   history -c
2   history 
3   handler -H 127.0.0.1 -P 4444 -p linux/x64/meterpreter_reverse_tcp 
.
.
.
18  history 
msf6 > handler -H 127.0.0.1 -P 4444 -p linux/x64/meterpreter_reverse_tcp 
[*] Payload handler running as background job 0.

[!] You are binding to a loopback address by setting LHOST to 127.0.0.1. Did you want ReverseListenerBindAddress?
msf6 > [*] Started reverse TCP handler on 127.0.0.1:4444 
[*] Meterpreter session 1 opened (127.0.0.1:4444 -> 127.0.0.1:39144) at 2021-04-25 19:24:57 +0530

msf6 > sessions -i 1 
[*] Starting interaction with 1...
```
No history in meterpreter when interacting for the first time
```
meterpreter > getuid 
Server username: pingport80 @ kali (uid=1000, gid=1000, euid=1000, egid=1000)
meterpreter > getpid 
Current pid: 10119
meterpreter > webcam_list 
1: USB2.0 HD UVC WebCam: USB2.0 HD
2: USB2.0 HD UVC WebCam: USB2.0 HD
meterpreter > background 
[*] Backgrounding session 1...
msf6 > history       # No meterpreter History should come
1   history -c
2   history 
3   handler -H 127.0.0.1 -P 4444 -p linux/x64/meterpreter_reverse_tcp 
.
.
.
19  handler -H 127.0.0.1 -P 4444 -p linux/x64/meterpreter_reverse_tcp 
20  sessions -i 1 
21  history 
msf6 > session -i 1
[*] Starting interaction with 1...

meterpreter >  # check with up-arrows if meterpreter history is showing correctly
```
### Verification Steps For Pry history
```
msf6 > history 
1  history -c
2  jobs 
3  sessions 
4  help 
5  id
6  help 
7  history 
msf6 > pry
[*] Starting Pry shell...
[*] You are in the "framework" object

[1] pry(#<Msf::Framework>)> history # No history will appear as this is the first interaction.

[2] pry(#<Msf::Framework>)> ls Rex
constants: 
.
.          
[3] pry(#<Msf::Framework>)> ls Rex::Ui
constants: Interactive  Output  Subscriber  Text
[4] pry(#<Msf::Framework>)> ls Rex::Ui::Text
constants: BidirectionalPipe  DispatcherShell  Input  IrbShell  Output  PseudoShell  Resource  Shell
[5] pry(#<Msf::Framework>)> history
1: history
2: ls Rex
3: ls Rex::Ui
4: ls Rex::Ui::Text
[6] pry(#<Msf::Framework>)> exit
msf6 > history 
1  history -c
2  jobs 
3  sessions 
4  help 
5  history 
6  pry
7  history 
msf6 > 
```
### History Files
The above verification was just for `readline` please check the history files also. 1) `~/.msf4/history` 2) `~/.msf4/meterpreter_history` 3) `~/.msf4/pry_history`
```
pingport80@kali:~/rapid7/metasploit-framework$ cat ~/.msf4/history 
history -c
history 
.
.
sessions -i 1 
history

pingport80@kali:~/rapid7/metasploit-framework$ cat ~/.msf4/met_history 
getuid 
getpid 
webcam_list 
background

pingport80@kali:~/rapid7/metasploit-framework$ cat ~/.msf4/pry_history 
history
ls Rex
ls Rex::Ui
ls Rex::Ui::Text
history
exit
pingport80@kali:~/rapid7/metasploit-framework$
```
